File: android/jsbridge/src/main/java/net/kibotu/jsbridge/commands/tracking/TrackEventCommand.kt
package net.kibotu.jsbridge.commands.tracking

import net.kibotu.jsbridge.commands.BridgeCommand
import net.kibotu.jsbridge.commands.utils.BridgeParsingUtils
import net.kibotu.jsbridge.commands.utils.BridgeResponseUtils
import timber.log.Timber

/**
 * Tracks analytics events (fire-and-forget).
 *
 * Enables unified analytics across native and web flows. Web developers can
 * track user behavior without native code changes, and events go through the
 * same native analytics pipeline.
 *
 * Fire-and-forget: returns null so no response is sent back to web.
 * Analytics failures should never block user actions.
 */
class TrackEventCommand : BridgeCommand {

    override val action = "trackEvent"

    override suspend fun handle(content: Any?): Any? {
        val event = BridgeParsingUtils.parseString(content, "event")
        if (event.isEmpty()) {
            return BridgeResponseUtils.createErrorResponse(
                "INVALID_PARAMETER",
                "Missing 'event' parameter"
            )
        }

        val params = BridgeParsingUtils.parseMap(content, "params")

        Timber.i("[Bridge] Track event: $event with params: $params")

        // TODO: Forward to your analytics service (e.g. Firebase Analytics)

        return null
    }
}

File: ios/JSBridge/Sources/JSBridge/Commands/tracking/TrackEventCommand.swift
import Foundation
import Orchard

/// Command for tracking analytics events (fire-and-forget)
public final class TrackEventCommand: BridgeCommand {
    public let action = "trackEvent"

    public init() {}

    @MainActor
    public func handle(content: [String: Any]?) async throws -> [String: Any]? {
        guard let event = content?["event"] as? String else {
            throw BridgeError.invalidParameter("event")
        }
        
        let params = content?["params"] as? [String: Any] ?? [:]
        
        let trackingEvent = BridgeTrackingEvent(name: event, parameters: params)
        Orchard.v("\(trackingEvent)")
        Orchard.v("[Bridge] Track event: \(event) with params: \(params)")
        
        return nil
    }
}

private struct BridgeTrackingEvent: TrackingEvent {
    let name: String
    let parameters: [String: Any]
}

public protocol TrackingEvent {
    var name: String { get }
    var parameters: [String: Any] { get }
}
